basic_pipeline/1_Introduction.md

```diff
@@ -0,0 +1,55 @@
+However, due to network limitations each frame sent by the peers is fragmented since only a few chars can be sent at the same time - that means, a single frame is sent as a sequence of **packets**.
+Each packet consists of the *header* (describing where the particular bunch of characters it transports should be put) and the *body* - aforementioned  bunch of characters.
+Below you can see an exemplary frame sent by one peer to the other. It gets fragmented into multiple packets, which later on are sent via the network (during that process their order will probably get messed). On the second peer's side, the packets get assembled into the original frame.
+![Example Chat](assets/images/example_chat.drawio.png)
+## Packet format
+Here is what each packet looks like:
+```
+[seq:<sequence_id>][frameid:<frame_id>][timestamp:<timestamp>]<text>
+```
+where:
++ sequence_id - the ordering number of the packet (relative to each of the peers). It reflects the position of a particular packet in the packets' stream.
++ frame_id - the identifier which consists of the number of the frame to which the body of a given packet belongs, optionally followed by a single **'e'** character (meaning that the packet is the **e**nding packet of the frame). Note that frames are numbered relatively to each peer in that conversation and that frame_id does not describe the global order of the frames in the final file.
++ timestamp - a number indicating a time at which a given sentence was said. Timestamp describes the order of the frames from both peers. 
++ text - the proper body of the packet, in our case - a bunch of characters which could be sent in a single packet.
+
+## Packets generator
+We have equipped you with the tool which produces the packets in the format described previously, based on the input conversation. You can use it as a mix task, by typing:
+```
+mix generate_input --packetsPerFrame <packets_per_frame> <input_file_path>
+```
+where: 
++ *input_file_path* is a path to the file which contains the text of the conversation. The input file consists of multiple lines of the following format:
+```
+<speaker id>: <text>
+```
++ *packets_per_frame* is a number of packets on which each frame will be dismentled
+
+Based on the input file content, that command will create multiple files, `<input_file_path>.<speaker id>.<input file format>` . For instance: `input.txt` with input from speakers `A` and `B` will be split into `input.A.txt` and `input.B.txt` files.
+The first file will contain the shuffled list of packets made from the `A` speaker's lines from the input file and the second file will contain a shuffled list of packets made out of the `B` speaker's lines of the input file. That `shuffle` is a way to simulate the imperfectness of the network - in real-life scenario, the order in which the packets are received is not always the same as the order in which they were sent.
```

Comment:
Maybe a link to some source explaining that in UDP packets can arrive out of order?

Reply:
Well, I was wondering about putting a link to an article which would describe some features of UDP, perhaps comparing it with TCP...but I have found [this](https://www.openmymind.net/How-Unreliable-Is-UDP/), which is a nice test - how about attaching it?

Reply:
done